### PR TITLE
FEAT: Add `delete_action`

### PR DESCRIPTION
**Pull request by Yuya-Furusawa**
I added `delete_action` function for `Player` and `NormalFormGame` instances in normal_form_game.jl
Also, add the test code of this function in test_normal_form_game.jl, and fix Games.jl

Close(#96)

### CODE REVIEW

**Comment by Yuya-Furusawa** [comment]
I will fix the error and send again.

File: src/normal_form_game.jl [hunk 1]
```diff
@@ -53,6 +53,75 @@ function Base.show(io::IO, player::Player)
...
+#Reference
+
+https://stackoverflow.com/questions/53637511/how-to-delete-the-specific-row-of-n-dimensional-array-in-julia
+"""
+
+function delete_action(player::Player{N},
+                  action::AbstractVector{<:PureAction}, player_idx::Int) where N
+    if -N <= player_idx <= 0
+      player_idx = player_idx + N
+    end
```
Comment (oyamad) [inline src/normal_form_game.jl]: Do not allow negative indexing in Julia. (The Python version just followed NumPy's convention.)

Comment (Yuya-Furusawa) [reply src/normal_form_game.jl]: `players_new  = [delete_action(player, action, player_idx-i+1) for (i, player) in enumerate(g.players)]` in the `delete_action` function for `NormalFormGame` may require `player_idx` negative, so I allowed negative indexing.

Should  I write another code which does not make index nonpositive?

Comment (oyamad) [reply src/normal_form_game.jl]: `player_idx` should be restricted in {1, ..., N}.

**Comment by Yuya-Furusawa** [comment]
I fixed them.

File: src/normal_form_game.jl [hunk 2]
```diff
@@ -53,6 +53,42 @@ function Base.show(io::IO, player::Player)
...
+    delete_action(player, action, player_idx)
+
+Return a new Player instance with the actions specified by
+`action` deleted from the action set of the player specified by
+`player_idx`.
+
+# Arguments
+
+- `player::Player` : Player instance.
+- `action::AbstractVector{<:PureAction}`: the actions to be deleted.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
- `action::Union{PureAction,AbstractVector{<:PureAction}}`: The action(s) to be
  deleted.
```

File: src/normal_form_game.jl [hunk 3]
```diff
@@ -53,6 +53,42 @@ function Base.show(io::IO, player::Player)
...
+- `player_idx::Integer` : Index of the player to delete actions for.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
- `player_idx::Integer` : Index of the player to delete action(s) for.
```

File: src/normal_form_game.jl [hunk 4]
```diff
@@ -53,6 +53,42 @@ function Base.show(io::IO, player::Player)
...
+
+# Returns
+
+- `::Player` : `Player` instance with the actions deleted as specified.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
- `::Player` : `Player` instance with the action(s) deleted as specified.
```

File: src/normal_form_game.jl [hunk 5]
```diff
@@ -53,6 +53,42 @@ function Base.show(io::IO, player::Player)
...
+
+#Reference
+
+https://stackoverflow.com/questions/53637511/how-to-delete-the-specific-row-of-n-dimensional-array-in-julia
```
Comment (oyamad) [inline src/normal_form_game.jl]: Do not put this in the docstring (this is just for internal implementation).

Just make a comment outside the docstring.

File: src/normal_form_game.jl [hunk 6]
```diff
@@ -53,6 +53,42 @@ function Base.show(io::IO, player::Player)
...
+"""
+
+function delete_action(player::Player{N},
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
function delete_action(player::Player{N}, action::AbstractVector{<:PureAction},
                       player_idx::Integer) where N
```

File: src/normal_form_game.jl [hunk 7]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
            " ",
            split(string(typeof(g)), ".")[end])
 
+# delete_action
+
+"""
+    delete_action(g, action, player_idx)
+
+Return a new `NormalFormGame` instance with the actions
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
Return a new `NormalFormGame` instance with the action(s) specified by `action`
deleted from the action set of the player specified by `player_idx`.
```

File: src/normal_form_game.jl [hunk 8]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
...
+specified by `action` deleted from the action set of the player
+specified by `player_idx`.
+
+# Arguments
+
+- `g::NormalFormGame` : `NormalFormGame` instance.
+- `action::AbstractVector{<:PureAction}` : the actions to be deleted.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
- `action::Union{PureAction,AbstractVector{<:PureAction}}` : The action(s) to be
  deleted.
```

File: src/normal_form_game.jl [hunk 9]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
...
+- `player_idx::Integer` : index of the player to delete action for.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
- `player_idx::Integer` : Index of the player to delete action(s) for.
```

File: src/normal_form_game.jl [hunk 10]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
...
+
+# Returns
+
+- `::NormalFormGame` : `NormalFormGame` instance with the actions deleted as specified.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
- `::NormalFormGame` : `NormalFormGame` instance with the action(s) deleted as
  specified.
```

File: src/normal_form_game.jl [hunk 11]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
...
+"""
+
+function delete_action(g::NormalFormGame{N},
```
Comment (oyamad) [inline src/normal_form_game.jl]: Indent:

```suggestion
function delete_action(g::NormalFormGame{N},
                       action::AbstractVector{<:PureAction},
                       player_idx::Integer) where N
```

File: src/normal_form_game.jl [hunk 12]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
...
+                        action::AbstractVector{<:PureAction},
+                        player_idx::Integer) where N
+  players_new  = [delete_action(player, action,
```
Comment (oyamad) [inline src/normal_form_game.jl]: Indent with 2 spaces.

File: src/normal_form_game.jl [hunk 13]
```diff
@@ -499,6 +535,39 @@ Base.summary(g::NormalFormGame) =
...
+                  player_idx-i+1>0 ? player_idx-i+1 : player_idx-i+1+N)
+                  for (i, player) in enumerate(g.players)]
+  return NormalFormGame(players_new)
```
Comment (oyamad) [inline src/normal_form_game.jl]: Ditto.

File: test/test_normal_form_game.jl [hunk 14]
```diff
@@ -67,6 +67,16 @@ using CDDLib
                        split(string(typeof(player)), ".")[end])
     end
 
+    @testset "Tests on delete_action for Player" begin
+        shapley_game = [0 1 0; 0 0 1; 1 0 0]
+        player = @inferred Player(shapley_game)
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Remove `@inferred`.

File: test/test_normal_form_game.jl [hunk 15]
```diff
@@ -141,6 +151,42 @@ using CDDLib
...
 
+    @testset "Tests on delete_action for NormalFormGame" begin
+        shapley_game = Array{Int}(undef, 3, 3, 2)
+        shapley_game[:, 1, 1] = [0, 0, 1]
+        shapley_game[:, 2, 1] = [1, 0, 0]
+        shapley_game[:, 3, 1] = [0, 1, 0]
+        shapley_game[:, 1, 2] = [0, 1, 0]
+        shapley_game[:, 2, 2] = [0, 0, 1]
+        shapley_game[:, 3, 2] = [1, 0, 0]
+        g = @inferred NormalFormGame(shapley_game)
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Remove `@inferred`.

File: test/test_normal_form_game.jl [hunk 16]
```diff
@@ -67,6 +67,16 @@ using CDDLib
                        split(string(typeof(player)), ".")[end])
     end
 
+    @testset "Tests on delete_action for Player" begin
+        shapley_game = [0 1 0; 0 0 1; 1 0 0]
+        player = @inferred Player(shapley_game)
+
+        @test delete_action(player, 1, 1).payoff_array ==
+            Player([0 0 1; 1 0 0]).payoff_array
+        @test delete_action(player, [1, 2], 1).payoff_array ==
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Add `@inferred player_new = delete_action(player, [1, 2], 1)`.

File: test/test_normal_form_game.jl [hunk 17]
```diff
@@ -141,6 +151,42 @@ using CDDLib
...
+
+        deleted_game_2 = Array{Int}(undef, 1, 3, 2)
+        deleted_game_2[1, 1, 1] = 1
+        deleted_game_2[1, 2, 1] = 0
+        deleted_game_2[1, 3, 1] = 0
+        deleted_game_2[1, 1, 2] = 0
+        deleted_game_2[1, 2, 2] = 1
+        deleted_game_2[1, 3, 2] = 0
+
+        @test delete_action(g, 1, 1).players[1].payoff_array ==
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Add `@inferred g_new = delete_action(g, 1, 1)`.

File: test/test_normal_form_game.jl [hunk 18]
```diff
@@ -141,6 +151,42 @@ using CDDLib
...
+            NormalFormGame(deleted_game_1).players[1].payoff_array
+        @test delete_action(g, 1, 1).players[2].payoff_array ==
+            NormalFormGame(deleted_game_1).players[2].payoff_array
+        @test delete_action(g, [1, 2], 1).players[1].payoff_array ==
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Add `@inferred g_new = delete_action(g, [1, 2], 1)`.

File: test/test_normal_form_game.jl [hunk 19]
```diff
@@ -67,6 +67,16 @@ using CDDLib
                        split(string(typeof(player)), ".")[end])
     end
 
+    @testset "Tests on delete_action for Player" begin
+        shapley_game = [0 1 0; 0 0 1; 1 0 0]
+        player = @inferred Player(shapley_game)
+
+        @test delete_action(player, 1, 1).payoff_array ==
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Add `@inferred player_new = delete_action(player, 1, 1)`.

File: src/normal_form_game.jl (hunk 6 above)
Comment (oyamad) [reply src/normal_form_game.jl]: Also, set a default value for `player_idx`:

```jl
                       player_idx::Integer=0) where N
```

**Comment by Yuya-Furusawa** [comment]
I fixed them again.

File: src/normal_form_game.jl [hunk 20]
```diff
@@ -53,6 +53,61 @@ function Base.show(io::IO, player::Player)
     Base.print_array(io, player.payoff_array)
 end
 
+# delete_action
+
+"""
+    delete_action(player, action, player_idx)
```
Comment (oyamad) [inline src/normal_form_game.jl]: Make the default value explicit:

```suggestion
    delete_action(player, action[, player_idx=1])
```

File: src/normal_form_game.jl [hunk 21]
```diff
@@ -53,6 +53,61 @@ function Base.show(io::IO, player::Player)
...
+ 3
+ 1
+
+"""
+
+# Reference
+# https://stackoverflow.com/questions/53637511/how-to-delete-the-specific-row-of-n-dimensional-array-in-julia
+
+function delete_action(player::Player{N,T}, action::AbstractVector{<:PureAction},
+                  player_idx::Integer=1) where {N,T}
```
Comment (oyamad) [inline src/normal_form_game.jl]: Indent.

File: src/normal_form_game.jl [hunk 22]
```diff
@@ -499,6 +554,40 @@ Base.summary(g::NormalFormGame) =
            " ",
            split(string(typeof(g)), ".")[end])
 
+# delete_action
+
+"""
+    delete_action(g, action, player_idx)
+
+Return a new `NormalFormGame` instance with the actions specified by `action`
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
Return a new `NormalFormGame` instance with the action(s) specified by `action`
```

File: src/normal_form_game.jl [hunk 23]
```diff
@@ -499,6 +554,40 @@ Base.summary(g::NormalFormGame) =
...
+- `action::Union{PureAction, AbstractVector{<:PureAction}}` : The action(s) to be
+  deleted.
+- `player_idx::Integer` : Index of the player to delete action(s) for.
+
+# Returns
+
+- `::NormalFormGame` : `NormalFormGame` instance with the action(s) deleted as
+  specified.
+"""
+
```
Comment (oyamad) [inline src/normal_form_game.jl]: Remove this blank line.

File: src/normal_form_game.jl [hunk 24]
```diff
@@ -53,6 +53,61 @@ function Base.show(io::IO, player::Player)
...
+ 3  0
+ 0  3
+julia> player_delete2 = delete_action(player, 1, 2)
+julia> player_delete2.payoff_array
+3×1 Array{Int64,2}:
+ 0
+ 3
+ 1
+
+"""
```
Comment (oyamad) [inline src/normal_form_game.jl]: Either

* remove the 4 lines between `"""` and `function delete_action(...)` (and move the link to stackoverflow to somewhere else); or

* surround `""" ... """` with `@doc` and `delete_action`:

  ```jl
  @doc """

  ...

  """ delete_action
  ```

  (Search for `@doc` within this file for a similar case.)

File: src/normal_form_game.jl [hunk 25]
```diff
@@ -53,6 +53,61 @@ function Base.show(io::IO, player::Player)
...
+julia> player_delete1.payoff_array
+2×2 Array{Int64,2}:
+ 3  0
+ 0  3
+julia> player_delete2 = delete_action(player, 1, 2)
+julia> player_delete2.payoff_array
+3×1 Array{Int64,2}:
+ 0
+ 3
+ 1
```
Comment (oyamad) [inline src/normal_form_game.jl]: Exactly copy-paste the outome!

```jl
julia> player = Player([3 0; 0 3; 1 1])
3×2 Player{2,Int64}:
 3  0
 0  3
 1  1

julia> delete_action(player, 3)
2×2 Player{2,Int64}:
 3  0
 0  3

julia> delete_action(player, 1, 2)
3×1 Player{2,Int64}:
 0
 3
 1
```

File: src/normal_form_game.jl [hunk 26]
```diff
@@ -499,6 +552,39 @@ Base.summary(g::NormalFormGame) =
...
+- `player_idx::Integer` : Index of the player to delete action(s) for.
+
+# Returns
+
+- `::NormalFormGame` : `NormalFormGame` instance with the action(s) deleted as
+  specified.
+"""
+function delete_action(g::NormalFormGame{N},
+                       action::AbstractVector{<:PureAction},
+                       player_idx::Integer=1) where N
```
Comment (oyamad) [inline src/normal_form_game.jl]: Do not set a default for `player_idx` in this method for `NormalFormGame`. (Player 1 is not special in any way, is it?)

**Comment by oyamad** [comment]
A new version (v1.1.1) of `CMake.jl` has just been released. Hopefully the next time it will build fine.

File: src/normal_form_game.jl [hunk 27]
```diff
@@ -499,6 +552,39 @@ Base.summary(g::NormalFormGame) =
...
+function delete_action(g::NormalFormGame{N},
+                       action::AbstractVector{<:PureAction},
+                       player_idx::Integer) where N
+    players_new  = [delete_action(player, action,
+                    player_idx-i+1>0 ? player_idx-i+1 : player_idx-i+1+N)
+                    for (i, player) in enumerate(g.players)]
+    return NormalFormGame(players_new)
+end
+
+delete_action(g::NormalFormGame, action::PureAction, player_idx::Integer=1) =
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
delete_action(g::NormalFormGame, action::PureAction, player_idx::Integer) =
```

**Comment by Yuya-Furusawa** [comment]
Sorry, I was confused. I just fixed them again and again.

**Pull request merged**